### PR TITLE
Add FuseConcatPass to eliminate redundant concat ops (#18827)

### PR DESCRIPTION
**Pull request by ryan-monroe**
Summary:

Adds an FX-level pass that eliminates concat ops which can be proven structurally redundant before the TOSA backend / Vela compiler ever sees them. In the Gen2 Executorch ARM / Ethos-U stack, `torch.cat` lowers to TOSA `CONCAT`, which Vela converts to N `MemoryCopy` ops — real DMA on the NPU. Catching the obvious cases up front keeps the TOSA flatbuffer fed to Vela smaller, keeps debug graphs honest, and provides defensive coverage on TOSA targets where Vela's own scheduler doesn't run (e.g., the VGF backend).

Five rewrite patterns are handled (inspired by Espresso's `bolt/nn/espresso/transforms/remove_nops.py`):

1. **Single-input concat**: `cat([x], dim) ≡ x` — replace cat with x.
2. **Concat-then-slice (exact)**: `cat([a, b, ...], dim)` feeding a `slice_copy` that extracts exactly one original input — replace the slice with the corresponding cat input directly.
3. **Slice-then-concat (full)**: `cat([slice(x, d, s0, e0), slice(x, d, s1, e1), ...], dim)` reconstructing x exactly (contiguous slices covering the full source dimension) — replace cat with x.
4. **Concat-then-sub-slice**: a `slice_copy` whose range falls entirely within one cat input — replace with an adjusted slice on that input directly.
5. **Slice-then-concat (partial)**: contiguous slices of the same tensor concatenated back but covering only a sub-range of the source — replace with a single slice on the source.

Authored with Claude.

Differential Revision: D97667069

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

File: backends/arm/_passes/fuse_concat_pass.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        _, s_start, s_end, _ = _slice_params(inp, source_dim_size)
+        if s_start != expected_start:
+            return None
+        expected_start = s_end
+
+    # expected_start is now the end of the last slice
+    return first_start, expected_start
+
+
+class FuseConcatPass(ArmPass):
```
Comment (Ninja91) [inline backends/arm/_passes/fuse_concat_pass.py:171]: @gggekov Here's another optimization pass we plan to add. Please comment if this will impact downstream passes/regor transforms

Comment (gggekov) [reply backends/arm/_passes/fuse_concat_pass.py:171]: I recommend testing the passes using `EthosU85PipelineINT` or `EthosU55PipelineINT` in order to know if the pass works with the current version of Vela.

**Comment by gggekov** [comment]
Helo @ryan-monroe , 
Thanks for the patch, great initiative to reduce the number of memcpy from CONCATs! 

I see in the unit tests, you only test with the `executorch.backends.arm.test.tester.test_pipeline.PassPipeline`. Given that the end metric you care for is number of CONCATs in the TOSA flatbuffer, wouldn't it make sense to test instead either with the `EthosU85PipelineINT` / `EthosU55PipelineINT` or `TosaPipelineINT` as these pipelines will generate a TOSA fb that is run on device? For counting the number of CONCAT ops in the TOSA fb, you can use the `count_tosa_ops` in `backends/arm/test/tester/test_pipeline.py`. See example how to do that here - https://github.com/pytorch/executorch/blob/main/backends/arm/test/misc/test_transpose_counts.py#L537

File: backends/arm/_passes/fuse_concat_pass.py (hunk 1 above)
Comment (gggekov) [reply backends/arm/_passes/fuse_concat_pass.py:171]: Overall, for me it makes sense for the optimization of some CONCATs to be done in ExecuTorch rather than Vela.

File: backends/arm/_passes/fuse_concat_pass.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    """Extract (dim, start, end, step) from a slice_copy node.
+
+    ``dim`` is normalized to a positive index. ``end`` is clamped to
+    ``dim_size`` (the size of the source tensor along the slice dimension).
+
+    """
+    rank = len(get_first_fake_tensor(node).shape)
+    dim = _int_arg(node, 1, 0)
+    dim = (dim + rank) % rank
+    start = _int_arg(node, 2, 0)
```
Comment (oscarandersson8218) [inline backends/arm/_passes/fuse_concat_pass.py:43]: Don't you need to normalize both start and end as these can be negative?

Comment (Ninja91) [reply backends/arm/_passes/fuse_concat_pass.py:43]: Would this be negative even when we export with static shapes? @oscarandersson8218

Comment (oscarandersson8218) [reply backends/arm/_passes/fuse_concat_pass.py:43]: Yes, I would think so as you can do `x[-3:3]`.

**Review by digantdesai — CHANGES_REQUESTED** [review]
Review automatically exported from Phabricator review in Meta.

File: backends/arm/_passes/fuse_concat_pass.py (hunk 2 above)
Comment (ryan-monroe) [reply backends/arm/_passes/fuse_concat_pass.py:43]: Addressed on all fronts :-) (or so I think!!)

Thanks for the feedback!

Also, thanks for the patience as I navigate this noob PR!

File: backends/arm/test/passes/test_fuse_concat_pass.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+    """
+    from executorch.backends.arm._passes.fuse_concat_pass import (
+        _find_common_slice_source,
+    )
+
+    assert _find_common_slice_source([None, None], cat_dim=0, dim_size=10) is None
+    assert _find_common_slice_source(["x", "y"], cat_dim=0, dim_size=10) is None
+
+
+# End-to-end Ethos-U pipeline test models. The PassPipeline modules above wire
```
Comment (gggekov) [inline backends/arm/test/passes/test_fuse_concat_pass.py:335]: Exactly - the PassPipeline only tests one pass in isolation. 

In reality, when you have a `torch.nn.Module` and call `to_edge_transform_and_lower`, the graph runs through 100+ passes which is why I vote to add more tests via the `EthosU85PipelineINT`/ `EthosU55PipelineINT` .  With `EthosU85PipelineINT`/ `EthosU55PipelineINT`, you are doing a holistic test with all the passes(including anything before and after the fuse_concat_pass) that will be run on the graph.

**Review by gggekov — COMMENTED** [review]
Thank you for updating the PR! 
I see there is a lint error- could you fix it and run through the CI ?
